### PR TITLE
refactor: extract Topology class

### PR DESCRIPTION
**Pull request by redeboer**
Improvements to the `topology` module in order to distinguish the different initialize stages of `StateTransitionGraph`s:

* Extracted `Topology` base class (no node/edge properties) from `StateTransitionGraph`
* Nodes have become a `set`
* Protected nodes and edges of `Topology`
* Type-templated edge properties of `StateTransitionGraph`
* `Edge` has become a `dataclass`

_Rebased on #247_

### CODE REVIEW

File: expertsystem/state/particle.py [hunk 1]
```diff
@@ -518,7 +518,7 @@ def assert_number_of_states(
     return new_graphs
 
 
-def __safe_set_spin_projections(
+def _safe_set_spin_projections(
```
Comment (spflueger) [inline expertsystem/state/particle.py]: why is that needed?

File: tests/unit/topology/test_graph.py [hunk 2]
```diff
@@ -58,9 +64,36 @@ def dummy_topology():
...
+        (["J/psi(1S)"], ["gamma", "pi0", "pi0"]),
+        (["J/psi(1S)"], ["K+", "K-", "pi+", "pi-"]),
+        (["e+", "e-"], ["gamma", "pi-", "pi+"]),
+        (["e+", "e-"], ["K+", "K-", "pi+", "pi-"]),
+    ],
+)
+def test_permutations(
+    initial_state, final_state, dummy_topology, particle_database
 ):
+    initial_state_with_spins = _safe_set_spin_projections(
```
Comment (spflueger) [inline tests/unit/topology/test_graph.py:89]: Ok, that's downside in the commit based reviews. This answers my question from before. However I'm questioning the `particle` API now. So concretely can the API be changed so that new function can provide this functionality (combined) and calls to  `_safe_set_spin_projection` can be avoided?

File: expertsystem/state/particle.py [hunk 3]
```diff
@@ -542,6 +545,23 @@ def safe_set_spin_projections(
     ]
 
 
+def generate_outer_edge_permutations(
+    topology: StateTransitionGraph,
+    initial_state: Sequence[StateWithSpins],
+    final_state: Sequence[StateWithSpins],
+) -> Generator[Dict[int, StateWithSpins], None, None]:
```
Comment (spflueger) [inline expertsystem/state/particle.py:550]: Does this function does not need the `StateWithSpins` right? In other words a particle name or pid would suffice. This connects to the question below

File: expertsystem/topology.py [hunk 4]
```diff
@@ -41,22 +41,23 @@ def __eq__(self, other: object) -> bool:
...
+    ) -> None:
         self.__nodes: Set[int] = set()
         self.edges: Dict[int, Edge] = {}
-        self.node_props: Dict[int, dict] = {}
-        self.edge_props: Dict[int, dict] = {}
-        self.graph_element_properties_comparator: Optional[Callable] = None
+        if nodes is not None:
+            self.__nodes = set(nodes)
+        if edges is not None:
+            self.edges = edges
```
Comment (spflueger) [inline expertsystem/topology.py]: This unsafe as the edge could point to non existing nodes etc

File: expertsystem/topology.py [hunk 5]
```diff
@@ -41,22 +41,23 @@ def __eq__(self, other: object) -> bool:
         raise NotImplementedError
 
 
-class StateTransitionGraph:
-    """Graph class that contains edges and nodes.
+class Topology:
+    """Directed graph for a state transition.
```
Comment (spflueger) [inline expertsystem/topology.py:37]: The thing is that this is not a normal graph of graph theory (CS). I think some comment explaining the difference would be good (maybe even a reference to feynman graphs)

File: expertsystem/topology.py [hunk 6]
```diff
@@ -22,23 +23,12 @@
...
-
-    def __eq__(self, other: object) -> bool:
-        if isinstance(other, Edge):
-            return (
-                self.ending_node_id == other.ending_node_id
-                and self.originating_node_id == other.originating_node_id
-            )
-        raise NotImplementedError
+    ending_node_id: Optional[int] = None
+    originating_node_id: Optional[int] = None
```
Comment (spflueger) [inline expertsystem/topology.py:33]: Ideal would be to avoid having the state (None, None), which is def invalid

File: expertsystem/topology.py [hunk 7]
```diff
@@ -260,7 +262,10 @@ def swap_edges(self, edge_id1: int, edge_id2: int) -> None:
         self.__edges[edge_id1] = popped_edge_id2
 
 
-class StateTransitionGraph(Topology):
+EdgeType = TypeVar("EdgeType")
```
Comment (spflueger) [inline expertsystem/topology.py:265]: can be hidden from api

**Review by spflueger — COMMENTED** [review]
Nice job altogether. I have a few question/comments

File: expertsystem/state/particle.py (hunk 1 above)
Comment (redeboer) [reply expertsystem/state/particle.py]: It is called in `test_graph.py`

File: expertsystem/state/particle.py (hunk 3 above)
Comment (redeboer) [reply expertsystem/state/particle.py:550]: Oh oops, this commit actually belongs to #245

Comment (redeboer) [reply expertsystem/state/particle.py:550]: I moved the commit to #245. But still, this `generate_outer_edge_permutations` uses the output of `_safe_set_spin_projections`, which is why the properties are `StateWithSpins` at this stage. Can also make it more abstract by using `Any`, because internally this function does not need `StateWithSpins`.

File: tests/unit/topology/test_graph.py (hunk 2 above)
Comment (redeboer) [reply tests/unit/topology/test_graph.py:89]: By new function, do you mean that `_safe_set_spin_projections` now takes a list? I'm not sure how `_safe_set_spin_projection can be avoided, because we want to allow the user to not specify spin projections in case all allowed spin projections should be considered.

File: expertsystem/topology.py (hunk 6 above)
Comment (redeboer) [reply expertsystem/topology.py:33]: Currently not possible because of the way in which `Topology` is built up elsewhere in the system through `add_edge`:

https://github.com/ComPWA/expertsystem/blob/5bca7a79be63fe5f18b2c10e151dce66b3f6dd62/expertsystem/topology.py#L91

File: expertsystem/topology.py (hunk 5 above)
Comment (redeboer) [reply expertsystem/topology.py]: 5bca7a79be63fe5f18b2c10e151dce66b3f6dd62

File: expertsystem/topology.py (hunk 7 above)
Comment (redeboer) [reply expertsystem/topology.py]: ae844f3e

File: tests/unit/topology/test_graph.py (hunk 2 above)
Comment (spflueger) [reply tests/unit/topology/test_graph.py]: Well the calls made here are really similar to what the `initialize_graphs` function is doing internally. So I'm saying, maybe instead `initialize_graphs` could be called (or some other new function, which would be a subpart of initialize graphs)

File: expertsystem/state/particle.py (hunk 3 above)
Comment (spflueger) [reply expertsystem/state/particle.py]: making it more abstract sounds good to me

File: tests/unit/topology/test_graph.py (hunk 2 above)
Comment (redeboer) [reply tests/unit/topology/test_graph.py]: > So I'm saying, maybe instead initialize_graphs could be called (or some other new function, which would be a subpart of initialize graphs)

I don't understand. `_safe_set_spin_projections` is called from `initialize_graphs` itself.

File: expertsystem/topology.py (hunk 6 above)
Comment (spflueger) [reply expertsystem/topology.py:33]: Ok, lets put that on the Todolist. Lets open a new Epic issue for that

Comment (redeboer) [reply expertsystem/topology.py:33]: Should that really be an epic?

Comment (spflueger) [reply expertsystem/topology.py:33]: those two classes are still quite unsafe

Comment (redeboer) [reply expertsystem/topology.py:33]: Yes that's true. I'm currently working on the `Topology.verify` method, to address at least some of those issues.

Comment (spflueger) [reply expertsystem/topology.py:33]: We can also make the way of building an invalid graph impossible, by changing the interface on how to add edges and nodes etc. For example these rules are quite solid:
- an Edge has to be attached to at least one node
- a new Node has to be attached to at least one edge (except the first node/seed node)

File: tests/unit/topology/test_graph.py (hunk 2 above)
Comment (spflueger) [reply tests/unit/topology/test_graph.py]: Hmm, why i can't see this change anymore? The old code seems to use it the way I was suggesting

File: expertsystem/topology.py (hunk 4 above)
Comment (redeboer) [reply expertsystem/topology.py]: #250

File: tests/unit/topology/test_graph.py (hunk 2 above)
Comment (redeboer) [reply tests/unit/topology/test_graph.py]: 7d780ec3

File: expertsystem/topology.py (hunk 6 above)
Comment (redeboer) [reply expertsystem/topology.py:33]: Discussion moved to #250

**Review by spflueger — APPROVED** [review]
